File: 323-number-of-connected-components-in-an-undirected-graph/number-of-connected-components-in-an-undirected-graph.py
from collections import defaultdict

class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:

        adj_list = defaultdict(list)
        for u, v in edges:
            adj_list[u].append(v)
            adj_list[v].append(u)


        visited = set()
        count = 0
        components = []

        def dfs(node, comp):
            # if node not in visited:
            visited.add(node)
            comp.add(node)
            
            for neighbor in adj_list[node]:
                if neighbor not in visited:
                    dfs(neighbor, comp)
            

        for i in range(n):
            if i not in visited:
                # print("I -> ", i)
                comp = set()
                dfs(i, comp)
                count += 1
                components.append(comp)
                
        # print("Count : ", count)
        print("components: ", components)
        return count

File: 323-number-of-connected-components-in-an-undirected-graph/number-of-connected-components-in-an-undirected-graph.java

import java.awt.Component;class Solution {
    public int countComponents(int n, int[][] edges) {

        List<List<Integer>> adjList = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            adjList.add(new ArrayList<>());
        }

        for (int[] edge : edges) {
            adjList.get(edge[0]).add(edge[1]);
            adjList.get(edge[1]).add(edge[0]);
        }
        
        boolean[] visited = new boolean[n];
        int count = 0;

        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                dfs(i, visited, adjList);
                count++;
            }
        }
        return count;
    }

    public void dfs(int node, boolean[] visited, List<List<Integer>> adjList) {

        visited[node] = true;

        for (int n : adjList.get(node)) {
            if (!visited[n]) {
                dfs(n, visited, adjList);
            }
        }
    }
}